### PR TITLE
Haiku patches

### PR DESCRIPTION
**Pull request by extrowerk**
Hi,

this is my first try to upstream the changes required for Haiku, i am open for any constructive critics.

Proof:
zALM

### CODE REVIEW

File: 3rdparty/bgfx/3rdparty/dear-imgui/widgets/file_list.inl [hunk 1]
```diff
@@ -32,6 +32,7 @@ namespace ImGui
 				}
 				else if (0 != ImStricmp(item->d_name, ".") )
 				{
+#ifndef __HAIKU__
```
Comment (extrowerk) [inline 3rdparty/bgfx/3rdparty/dear-imgui/widgets/file_list.inl]: No d_type on Haiku.

File: 3rdparty/bx/src/os.cpp [hunk 2]
```diff
@@ -46,10 +48,13 @@
 #		include <malloc.h> // mallinfo
 #	elif   BX_PLATFORM_LINUX     \
 		|| BX_PLATFORM_RPI       \
+		|| BX_PLATFORM_HAIKU  \
 		|| BX_PLATFORM_STEAMLINK
 #		include <stdio.h>  // fopen
 #		include <unistd.h> // syscall
+#ifndef __HAIKU__
 #		include <sys/syscall.h>
```
Comment (extrowerk) [inline 3rdparty/bx/src/os.cpp]: Haiku don't have this.

File: src/osd/modules/file/posixptty.cpp [hunk 3]
```diff
@@ -19,6 +19,10 @@
 #include <unistd.h>
 #include <stdlib.h>
 
+#ifdef __HAIKU__
+#define _BSD_SOURCE
```
Comment (extrowerk) [inline src/osd/modules/file/posixptty.cpp]: openpty is non-POSIX.

**Comment by extrowerk** [comment]
Any input?

**Comment by rb6502** [comment]
The vast majority of this patchset is actually against things which aren't MAME; GENie, BX, and dear imgui.  We'd like to see you at least attempt to upstream these past us so then we don't have to re-add Haiku support each time we resync with those sources.  GENie and BX live at:

https://github.com/bkaradzic/bx
https://github.com/bkaradzic/genie

And dear imgui is from:
https://github.com/ocornut/imgui

Thanks!  We do appreciate the effort :)

**Comment by extrowerk** [comment]
Tried to upstream the genie patches, but the maintainer won’t merge it as he cannot test it.
Will try the others.

**Pull request closed without merging**